Add monthly pricing test

taggon opened the pull request:
#### Changes proposed in this Pull Request

* Add the monthly pricing test to Calypso. This PR requires the server-side implementations.

#### Testing instructions

💡  For full and **highly-detailed test instructions**, please see: pau2Xa-2ak-p2

-- or --

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Make sure you're proxied.
* Sandbox `public-api.wordpress.com` which D51146-code is applied to.
* (added) D51640-code also needs to be placed on the sandbox.
* ~~Assign yourself to a test group of `monthly_pricing_test_phase_1` by following the instruction ( pau2Xa-251-p2#comment-7999 )~~
* Assigny yourself to the `treatment` group of `monthlyPricing` a/b test.
* Start the process of creating a new website with the `onboarding` flow( `/start/user` ).
* If you're in the `treatment` test group, the plans page should show monthly dotcom plans as below:
  <img width="893" alt="Create_a_site_—_WordPress_com" src="https://user-images.githubusercontent.com/212034/96158436-7b928e00-0f4e-11eb-9e50-06311af0c325.png">
* Pick a plan except for the free plan.
* Your site is newly created and you will see the checkout page that shows what you get with annual/biennial plans.
  <img width="811" alt="checkout_e28094_wordpress_com-2" src="https://user-images.githubusercontent.com/212034/96991514-81e2c480-1563-11eb-9cd4-c5dc52b73651.png">
  <img width="811" alt="checkout_e28094_wordpress_com-3" src="https://user-images.githubusercontent.com/212034/96991532-8909d280-1563-11eb-89bd-b071504861ab.png">

taggon committed:
Add monthly pricing test

taggon committed:
Make more progress on the plans page.

taggon committed:
Change the experiment id

taggon committed:
Fix lint errors / broken tests

taggon committed:
Update plans features for monthly pricing test

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.js:
```diff
@@ -113,12 +142,16 @@ function CheckoutSummaryFeaturesList() {
 	} else if ( hasPlanInCart && ! hasDomainsInCart ) {
 		refundText = translate( '30 day money back guarantee' );
+		if ( isMonthlyPricingTest && hasMonthlyPlan ) {
+			refundText = translate( '5 day money back guarantee' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( '4 day money back guarantee' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/monthly-pricing-abtest)

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.js:
```diff
@@ -201,21 +253,44 @@ function CheckoutSummaryPlanFeatures() {
+
+	if ( [ 'personal-bundle', 'personal-bundle-2y' ].includes( productSlug ) ) {
 		return [
+			isMonthlyPricingTest && translate( 'Live chat and email support' ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **33** times:
`translate( 'Email and live chat support' )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/monthly-pricing-abtest)

fditrapani commented:
This is coming along nicely @taggon. I know it's a work in progress but just wanted to share a couple points:

- For the toggles, can we add a hover state to the inactive tabs and increase the hover/clickable area:
![image](https://user-images.githubusercontent.com/6981253/96267810-44e06600-0f96-11eb-96e8-ac41977c63a3.png)
- Can we make the inactive tab text darker(Gray 80/#2C3338), it currently looks disabled. 
- Can we add a subtle animation to the popover message when you toggle to the monthly view to make it a little more noticeable? Something like this but smoother (I think some frames dropped from the recording):

![animate-in](https://user-images.githubusercontent.com/6981253/96281298-25056e00-0fa7-11eb-878c-34af52944f56.gif)


- Can we make the "INCLUDED WITH ANNUAL PLANS" text slightly bigger. I had it set to 10px in the mockups. 
- There was some back and forth with the free site card and we ultimately decided to remove the card and add a text link in the sub header:
![image](https://user-images.githubusercontent.com/6981253/96274238-4877eb00-0f9e-11eb-8ea3-b2b94e6bb3b6.png)

- And lastly, just wanted to mention the popup on the mobile view appearing under the tabs:
![image](https://user-images.githubusercontent.com/6981253/96281453-5847fd00-0fa7-11eb-94cc-b03e843866c0.png)

jessie-ross commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -536,29 +417,66 @@ export class PlansFeaturesMain extends Component {
+				<HappychatConnection />
+				<div className="plans-features-main__notice" />
+
+				<ExperimentActivator
+					name="monthly_pricing_test_phase_1"
+					isActive={ isEligibleForMonthlyPricing }
+				>
+					<DefaultVariation name="control">
+						<PlanTypeSelector
+							{ ...this.props }
+							basePlansPath={ basePlansPath }
+							isMonthlyPricingTest={ false }
+						/>
+						<FreePlanBanner { ...this.props } hidden={ shouldHideFreePlanBanner } />
+						{ this.getPlanFeatures( { isMonthlyPricingTest: false } ) }
+						{ this.renderProductsSelector() }
+					</DefaultVariation>
+					<Variation name="treatment">
+						<PlanTypeSelector
+							{ ...this.props }
+							basePlansPath={ basePlansPath }
+							isMonthlyPricingTest={ true }
+						/>
+						{ this.getPlanFeatures( { isMonthlyPricingTest: true } ) }
+						{ this.renderProductsSelector() }
+						<FreePlanBanner { ...this.props } hidden={ shouldHideFreePlanBanner } />
+					</Variation>
+				</ExperimentActivator>
```
Hi @taggon, I think something like this is more suitable:

```suggestion
				<Experiment
					name="monthly_pricing_test_phase_1"
				>
					<DefaultVariation name="control">
						<PlanTypeSelector
							{ ...this.props }
							basePlansPath={ basePlansPath }
							isMonthlyPricingTest={ false }
						/>
						<FreePlanBanner { ...this.props } hidden={ shouldHideFreePlanBanner } />
						{ this.getPlanFeatures( { isMonthlyPricingTest: false } ) }
						{ this.renderProductsSelector() }
					</DefaultVariation>
					<Variation name="treatment">
						<PlanTypeSelector
							{ ...this.props }
							basePlansPath={ basePlansPath }
							isMonthlyPricingTest={ isEligibleForMonthlyPricing }
						/>
						{ this.getPlanFeatures( { isMonthlyPricingTest: isEligibleForMonthlyPricing } ) }
						{ this.renderProductsSelector() }
						<FreePlanBanner { ...this.props } hidden={ shouldHideFreePlanBanner } />
					</Variation>
				</Experiment>
```

Here I have:

1. Changed from `ExperimentActivator` to `Experiment`. 

    It seems like possibly `ExperimentActivator` exists to avoid loading state, has that been an issue for you?

    It also might exist to avoid unnecessary assignment? An experiment assignment in calypso is cheap and since you will be using an "exposure event" to track when a user has been exposed to an experiment rather than assignment, when an assignment happens won't matter. If this is the case feel free to use the `Experiment` component as many times as you would like, it might help to think of it as just checking the assignment rather than causing an assignment.

2. Moved the eligibility logic within the `treatment` variation.

    This should help simplify things as the experimentation system shouldn't need to know about this, only the treatment case. It might even help to pass this further below in two spots rather than here, but I don't know this code well enough to say for sure.

taggon committed:
Update styles

taggon commented:
**Plans step**
* [x] Add hover state to the inactive tabs
* [ ] Increase the hover/clickable area in the tabs. (Can you give me a number?)
* [x] Darken the inactive tab text
* [ ] Animate popup messages.
* [x] Make "INCLUDED WITH ANNUAL PLANS" bigger.
* [x] Remove free site banner. Add a link to the subtext instead.
* [x] Show popups below the tabs on the mobile view.

cc @fditrapani

fditrapani commented:
Thanks for making those updates @taggon! The hover area looks good with what you have!

taggon committed:
Add an appearing animation to popup messages

taggon commented:
@fditrapani The popup messages now appear with animation.

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@taggon please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/929661). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

fditrapani commented:
Looks good @taggon! Thanks for making the update!

taggon committed:
Show monthly pricing specific UI on the checkout page

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-review.js:
```diff
@@ -121,6 +125,53 @@ function CouponFieldArea( {
+
+	let sublabel = translate( 'Monthly subscription' );
+	if ( billingMonths === 12 ) {
+		sublabel = translate( 'One year subscription' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'for a one year subscription' )` *ES Score: 8*

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.js:
```diff
@@ -113,12 +142,16 @@ function CheckoutSummaryFeaturesList() {
 	} else if ( hasPlanInCart && ! hasDomainsInCart ) {
 		refundText = translate( '30 day money back guarantee' );
+		if ( isMonthlyPricingTest && hasMonthlyPlan ) {
+			refundText = translate( '7 day money back guarantee' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( '4 day money back guarantee' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/monthly-pricing-abtest)

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.js:
```diff
@@ -471,13 +524,24 @@ function shouldLineItemBeShownWhenStepInactive( item ) {
 
 	if ( item.type === 'plan' && item.wpcom_meta?.months_per_bill_period > 1 ) {
+		if ( isMonthlyPricingTest ) {
+			return translate( '%(sublabel)s: %(monthlyPrice)s /month × %(monthsPerBillPeriod)s', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( '%(sublabel)s: %(monthlyPrice)s per month × %(monthsPerBillPeriod)s' )` *ES Score: 8*

taggon committed:
Display how much users save with annual/biennial plans

taggon commented:
Rebased.

taggon commented:
Something went wrong with the previous rebase. Rebased again.

taggon committed:
Fix failed test cases

fditrapani commented:
This is looking really good for Desktop @taggon. I noticed some weird behaviour on mobile:

![image](https://user-images.githubusercontent.com/6981253/96890680-de57be00-1455-11eb-9b48-6eb32b824b8d.png)

If you have annual selected, there's a big space between the tabs and the plans. Then when I switch over to monthly, the tool tip shows momentarily and then disappears. 

A small comment on the order of the features on the personal plan, it looks a little weird for the one feature that changes to be in the middle, can we move the free domain to the top of the list so support appears below it?

I'm also not seeing the pricing and active buttons for the plans — I'm guessing that might require another back-end diff?

sirbrillig commented:
Sorry I caused a conflict by moving the shopping-cart types to a package in https://github.com/Automattic/wp-calypso/pull/46532 but the fix should be fairly straightforward.

taggon commented:
@sirbrillig No problem! Thanks for letting me know that. :)

taggon committed:
Add missing plan features to the checkout page

taggon commented:
Rebased to resolve conflicts.

taggon committed:
Fix type errors

taggon commented:
> If you have annual selected, there's a big space between the tabs and the plans. Then when I switch over to monthly, the tool tip shows momentarily and then disappears.
Is the expected behavior moving the plans cards to the bottom when the monthly tab is selected?

taggon committed:
Fix the plans page design

taggon commented:
@fditrapani

_Fixed_ ~~A small comment on the order of the features on the personal plan, it looks a little weird for the one feature that changes to be in the middle, can we move the free domain to the top of the list so support appears below it?~~

_Fixed_ ~~If I pick a paid plan, I'm still seeing the message to create a free site. The copy should read "Choose a plan. Upgrade as you go."~~

> I'm also not seeing the pricing and active buttons for the plans — I'm guessing that might require another back-end diff?

There are only two diffs you should set up for this PR. You may need to set USD currency. See p1603335800039500-slack-CV2TX2PAN

> If you have annual selected, there's a big space between the tabs and the plans. Then when I switch over to monthly, the tool tip shows momentarily and then disappears.

Is the expected behavior moving the plans cards to the bottom when the monthly tab is selected?

fditrapani commented:
Thanks for the updates @taggon! 

> Is the expected behavior moving the plans cards to the bottom when the monthly tab is selected?

Yes, the plans will move down when you select the monthly plan. Here's a screenshot:
![image](https://user-images.githubusercontent.com/6981253/97014946-f8a6a000-1518-11eb-9ddc-3c37c09449c4.png)

I also got a chance to test checkout. The monthly toggle is working well. The only thing that's missing is the toggle at the bottom of the feature list to add the annual plan (if you have the monthly plan selected). This will require moving the support messaging.

![image](https://user-images.githubusercontent.com/6981253/97015145-35729700-1519-11eb-963f-b4a1e4c6e456.png)

taggon committed:
Remove spacing between tabs and plans when yearly plan is selected

taggon committed:
Add "switch to annual plan" CTA to order summary section for monthly plans

taggon commented:
@fditrapani 
1. The big spacing on the mobile view is now removed. It only appears when the monthly plan tab is selected.
2. Can you elaborate on the expected behavior when the _Switch to annual plan_ button is clicked? The button is added by the latest commit but it does nothing right now.

fditrapani commented:
Thanks @taggon 

> The big spacing on the mobile view is now removed. It only appears when the monthly plan tab is selected.

Nice! That looks better. There's still a little jump in the animation but otherwise looks great!

![plan](https://user-images.githubusercontent.com/6981253/97187917-52a5a080-1779-11eb-9ee6-b257ed90dfa5.gif)

> Can you elaborate on the expected behavior when the Switch to annual plan button is clicked? The button is added by the latest commit but it does nothing right now.

Clicking on that button should replace the monthly plan in your cart with the yearly one and as a result, the sidebar content should update. Here's a screenshot from my mockup with some slightly out of date copy for live chat on the personal plan:

![image](https://user-images.githubusercontent.com/6981253/97188143-926c8800-1779-11eb-9897-c3e4bcd83b01.png)


While I was looking at checkout, I noticed a couple little things on mobile:

1. The description for the yearly plan costs is going over the reason for the discount:
![image](https://user-images.githubusercontent.com/6981253/97188545-fee78700-1779-11eb-8f58-8bc16ac46824.png)

2. The support prompt in the order summary is sitting ontop of the Your Order title:
![image](https://user-images.githubusercontent.com/6981253/97188849-5ab21000-177a-11eb-8f9b-43d396da29f5.png)

Changing the position absolute to relative, and adding some padding, should fix that up. Here's how it should look:
![image](https://user-images.githubusercontent.com/6981253/97189208-b8465c80-177a-11eb-8e0c-3aa24abea606.png)

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5048137

Thank you @taggon for including a screenshot in the description! This is really helpful for our translators.

withinboredom commented on client/signup/steps/plans/index.jsx:
```diff
@@ -215,6 +246,22 @@ PlansStep.propTypes = {
+	const isEligibleForMonthlyPricing = 'onboarding' === props.flowName;
+
+	return (
+		<Experiment name="monthly_pricing_test_phase_1">
+			<DefaultVariation name="control">
+				<PlansStep { ...props } isMonthlyPricingTest={ false } />
+			</DefaultVariation>
+			<Variation name="treatment">
+				<PlansStep { ...props } isMonthlyPricingTest={ isEligibleForMonthlyPricing } />
+			</Variation>
+		</Experiment>
```
This could ruin the experiment.

The analysis is going to report the users who `isEligibleForMonthlyPricing == false` as being in the experiment and experiencing the treatment, but really they're experiencing the control. So they'll bias the experiment.

Instead, I propose the following:
```suggestion
		(!isEligibleForMonthlyPricing 
			? <PlanStep { ...props } isMonthlyPricingTest={ false } />
			: <Experiment name="monthly_pricing_test_phase_1">
			<DefaultVariation name="control">
				<PlansStep { ...props } isMonthlyPricingTest={ false } />
			</DefaultVariation>
			<Variation name="treatment">
				<PlansStep { ...props } isMonthlyPricingTest={ true } />
			</Variation>
		</Experiment>)
```

withinboredom replied:
Actually, I just realized we also put them in the experiment on the server-side, so you'll need to remove the server-side variation retrieval and leave eligibility specifically for the client, or remove eligibility being handled in the client.

taggon replied:
@withinboredom We no longer use the `<Experiment>` component. Instead, we query the experiments with `<QueryExperiments />` and pull the data using hooks.

withinboredom reviewed (changes requested):
The hook is probably minor (I'm not familiar with how the hook is used), but the `isEligibleForMonthlyPricing` def needs to be fixed before it goes live.

Edit, I also just noticed that we don't check eligibility elsewhere. That could def cause some issues with parts of the stack thinking they're in the experiment and other parts disagreeing. Dunno if that's an actual issue, but thought I'd point it out.

taggon committed:
ExPlat: Add hooks

taggon committed:
Exclude paid media from the test

taggon committed:
Update plans features

taggon commented:
Rebased.

taggon commented:
@fditrapani All fixed! The only remaining - the jumping popup is caused by the `<Popover />` component that has not changed by this PR. I'd like to keep it as a known problem for now because we still have many priorities.

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary.js:
```diff
@@ -157,42 +196,65 @@ function SupportText( { hasPlanInCart, isJetpackNotAtomic } ) {
+function CheckoutSummaryFeaturesListDomainItem( { domain, isMonthlyPricingTest, hasMonthlyPlan } ) {
 	const translate = useTranslate();
+	const bundledText = isMonthlyPricingTest
+		? translate( 'free for one year' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **32** times:
`translate( 'Free domain for one year' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/monthly-pricing-abtest)

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@taggon please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/935068). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

taggon committed:
Fix cart broken for non-test group

taggon committed:
Show '[PLAN NAME] monthly' on My Plan

withinboredom reviewed (approved):
Experiment implementation LGTM

withinboredom commented on client/signup/steps/plans/index.jsx:
```diff
@@ -250,17 +252,18 @@ PlansStep.propTypes = {
+		<>
+			<QueryExperiments />
+			{ ! isLoading && (
 				<PlansStep { ...props } isMonthlyPricingTest={ isEligibleForMonthlyPricing } />
```
I forgot to mention, it may be a good idea to add a prop to a tracks event (or fire a new one) that includes the status of `isEligibleForMonthlyPricing` which will help filter those out if this does end up biasing the test in some way. It's not a blocker, just a heads up; just in case.

aaronyan replied:
I like this suggestion, just slightly altered!

Consider a separate variable `isEligibleForMonthlyPricingExperiment` which checks for 
```
'onboarding' === props.flowName && ! isIgnoredAdSource()
```

And when true, then trigger a Tracks event like `calypso_experiment_exposure_eligible` with eventprop `experiment_name: monthly_pricing_test_phase_1`. You can then use that as your exposure event in your Abacus experiment analysis.

This will help ensure you are only analyzing over users that are eligible for your experiment. Without this particular exposure event, you may be diluting the experiment analysis by preventing some `treatment` participants from experiencing the monthly plans pricing.

aneeshd16 committed:
Fixes getStoreSlug for PLAN_ECOMMERCE_MONTHLY

taggon committed:
Move back to the old a/b testing framework

aneeshd16 committed:
Fixes issue where component was cast into string

aneeshd16 committed:
Merge branch 'add/monthly-pricing-abtest' of https://github.com/Automattic/wp-calypso into add/monthly-pricing-abtest

taggon committed:
Make "Switch to Annual plan" work

taggon committed:
Show always free domain and live chat

taggon committed:
Revert "ExPlat: Add hooks"

This reverts commit 1d67c6809beb44bd6cd81e348b80dc14535830a4.

taggon committed:
Revert the changes in checkout test suites

taggon committed:
Fix the domain item in the checkout summary

taggon committed:
Hide dotcom monthly plans for non test group

taggon committed:
Do not show crossed out price for control group

taggon committed:
Fix FreePlanBanner is not working

taggon committed:
Activate the test only when a dotcom plan exists in the cart

taggon committed:
Fix annual only label shows up to the control group

taggon committed:
Fix customerType selector generates wrong urls

taggon committed:
Fix monthly plan does not appear for the variant

taggon committed:
Fix broken unit tests

taggon committed:
Hide bundled domain ToS for monthly plans

taggon committed:
Hide "Switch to annual" for renewal items

jaswrks reviewed (approved):
- Tracks events LGTM. ✅ 
- Paid media exclusions LGTM. ✅ 
- Treatment flow in monthly test LGTM. ✅ 
- Control flow in monthly test LGTM. ✅ 
- Navigating Calypso UI w/ monthly plan LGTM. ✅ 

Things I did not test:
- Did not test all possible actions taken in Calypso w/ monthly plan.
- Only did a cursory review of the code changes, as these changes touch areas that need more Calypso experience than I have to offer you.

taggon committed:
Fix: eCommerce monthly shows annual plans in the plans page

taggon committed:
Exclude renewal items from the test

dzver committed:
Fix Upgrade to Business Nudge for Monthly Plans (#46852)

* The upgrade slug for monthly plan should be business-monthly

Providing an annual slug for a monthly plan broke the code that
identifies the upgrade path

* The 30-days are hardcoded

We would also want to use the existing translation until the new one
becomes available

* Adds test for maybeShowPlanBumpOffer with a monthly plan

Co-authored-by: Aneesh Devasthale <aneeshdevasthale@gmail.com>

Commit:
Merge branch 'master' into add/monthly-pricing-abtest

aneeshd16 committed:
Removes invalid params introduced in conflict fix

taggon committed:
Fix negative discount

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5048137).
